MSC4278: Media preview controls (#4278)
matrix-org/matrix-doc: proposals/4278-media-preview-controls.md

# MSC4278: Media preview controls

Matrix caters to a wide variety of networks configurations. Some networks are closed and the users have a high
degree of safety, and other networks are open to the public internet and are inherently less safe. This proposal
aims to give the user more universal control over the content they see on Matrix, starting with media previews.

Most graphical Matrix clients display media in the timeline of a room with a preview. Often these are delivered
without any consent prompt, at least by default. This presents a risk to users who may be joining the network
and do not have the same safety barriers that experienced users may have configured.

With that in mind, this MSC provides a simple switch for users and homeservers to configure a global media preview
preference.

For the purposes of this MSC, the term "undesirable content" refers to images that the user does not want to see,
be it illegal, abusive or otherwise unwanted.

## Proposal

This proposal introduces a new account data key, `m.media_preview_config`.

This key contains the following content.

```json
{
  "media_previews": "off|private|on",
  "invite_avatars": "off|on"
}

> ```

Author: My main concern is the power in this MSC is the simplicity, which makes it easier for clients to protect their users. We could instead have an array of "room types" and "on" / "off", but it becomes much harder to build a UI around that based on our previous attempts.

I've mentioned in the MSC that if there is an extension, then clients that do not support the new extension should treat the value as `off` for safety. 

My feeling is that we can and should consider more nuanced cased for previews but it should be pushed to the next MSC so that we can get this out and give users a coarse setting to protect themselves (as unfortunately this has already sat around far too long).

> ```

Reviewer: @Johennes switching `media_previews` to something like an array (e.g. `["private", "moderated", "public", ...]`) would allow for more future extensibility. But it does lead to potentially confusing combinations (a user sets the value to `["private", "moderated"]`. Should previews be shown in public, but moderated rooms?).

On the UI discussion, we currently have the following in Element:

<img width="569" height="384" alt="Image" src="https://github.com/user-attachments/assets/dda51631-8516-49d0-9769-79d8e202e0d6" />

You could switch this to checkboxes:

<img width="1526" height="1030" alt="Image" src="https://github.com/user-attachments/assets/e146fcc8-b296-4004-bda7-01e5de5cbef6" />

But the fact that "moderated" can overlap with "public" and "private" leads to it being difficult to predict what should happen in a given room.

Perhaps something like a slider view would be better:

<img width="1536" height="1024" alt="Image" src="https://github.com/user-attachments/assets/3ed37482-320f-43c5-b4e2-c4fb5e7cb738" />

And to support "private and moderated", you can define a new "private_moderated" (or just "moderated") value for `media_previews`. No array needed.

Personally I like the idea of each value opening up more rooms where previews would be shown by default. The simple string type for `media_previews` works in that case.

---

@Half-Shot I imagine you've already looked at this, but Discord takes a separate approach, allowing for lots of extensibility:

<img width="919" height="369" alt="Image" src="https://github.com/user-attachments/assets/b6cb7a7e-917c-433b-a929-cc59ff405cd8" />

I would model this as:

```json
{
  "msm": {
    "direct_messages": "blur",
    "direct_messages_friends": "block",
    "server_channels": "blur"
  },
  "gm": {
    "direct_messages": "blur",
    "direct_messages_friends": "block",
    "server_channels": "blur"
  },
}
```

Though this seems overkill for our use case, and I think the slider above is much more user-friendly (though the UI that Element has today is fine).

> ```

Author: We talked about extensibility in a sync today.

For what it's worth my feeling is that for minor changes such as new enums, the current behaviour in the spec is fine insofar that unknown enums are treated as `off` which means it fails-safe. As clients update, they can enjoy the complexity.

I feel that for larger extensions, we don't need to account for this in the current MSC. There are options such as creating a more featureful event type and using this as the fallback for older clients, but that's out of scope here.

### `media_previews`

`media_previews` refers to any media that may be automatically rendered in the context of a room without user
interaction, like a thumbnail.

Clients should show show or hide media depending upon this setting, on at least the following content types:

 - Media message types `m.image`, `m.video`, `m.file` which all can include thumbnail information or be thumbnailed
 by the home server.
 - The `m.sticker` event type.
 - Inline media via `img`.

>  - Any event which includes thumbnail information or may be thumbnailed by the home server.

Author: The EWeb implementation does, so we should!

required to be filtered by the same logic. A future MSC may explore how to handle user avatars.

### `invite_avatars`

`invite_avatars` refers to any **room** avatar rendered in the client, where the client has a `membership` of `invite`.

This may typically be rendered in the form of an invite dialog, or a room list preview of a room.

The avatar may refer to:

 - The `m.room.avatar` state event.
 - The inviting user's `avatar_url` from their profile (in the case of a DM).

In *all* cases where an avatar may be rendered for a room invite, the preview should be controlled by the setting.

### Property value

#### `off`

The client MUST NOT show any previews for any media in affected rooms. Clients SHOULD hide the media entirely,
behind a click-to-view prompt, or some other mechanism where a user is either prevented entirely or must consent
to see the media.

Users may individually consent to seeing media, for example by clicking on a prompt to show a preview.

If consent is given, the client SHOULD then track that consent and show the media again in the future.

#### `private`

Previews for media MAY be shown in "private" rooms without a prompt. A private room is any room where 

> the `m.room.join_rules` state exists AND `join_rule` key of this state is `invite`, `knock`, `restricted`, or `knock_restricted`.

Author: @clokep @turt2live I'm confused by this feedback, should I be changing the word "private" so as not to clash?

> the `m.room.join_rules` state exists AND `join_rule` key of this state is `invite`, `knock`, `restricted`, or `knock_restricted`.

Reviewer 2: I don't know the solution besides that this is defining a new term. That is probably ok if it doesn't conflict with future T&S or other plans.

> the `m.room.join_rules` state exists AND `join_rule` key of this state is `invite`, `knock`, `restricted`, or `knock_restricted`.

Reviewer 2: This might need @jimmackenzie inout also, but "private" means many things to different people. There's been a bunch of previous conversations around what a "private room" is.

I don't really want to block this on naming stuff, but this term is also pretty loaded.

From my POV I'm looking for input from other SCT members here.

> the `m.room.join_rules` state exists AND `join_rule` key of this state is `invite`, `knock`, `restricted`, or `knock_restricted`.

Reviewer 3: No changes required from my perspective - just flagging that the definition might change later.

> the `m.room.join_rules` state exists AND `join_rule` key of this state is `invite`, `knock`, `restricted`, or `knock_restricted`.

Reviewer 2: Another option could be to have an array of join rules to allow, but that trades flexibility for probably a worse UI of some sort. 🤔 I'm still a bit concerned about adding a new term to the spec that's overloaded.

If any other `join_rule` is set, or cannot be determined by the client then the assumption MUST be that the
room is public and previews MUST not be shown. Future join rules may be added to this list, but it's critical
that clients adopt a safety first approach here.

Note that this setting has no effect for `invite_avatars`. Avatars can only be `off` or `on` for all invites. 
Bad actors can easily send a DM to a user (which would pass the `private` check) containing unwanted
content.

#### `on`

> Media MAY be shown in any room without a prompt. 

Author: Updated 08ca88744090b9a1185bf5e7e46582b4ab24cd39

> Users may individually hide media, and this preference MUST be respected over any defaults defined in `m.media_preview_config`.

Author: I think it's still essential to allow clients to do this, but it's not a hard requirement.

> The mechanism for tracking opt-in/opt-out of media is left as an implementation detail.

A future MSC could de-implementation detail it, but for the moment this system is working out quite well for the Element clients.

> Users may individually hide media, and this preference MUST be respected over any defaults defined in `m.media_preview_config`.

Author: Resolving as a future addition for a future MSC.

This value is the **default** setting for both properties when no value is set, to keep with the previous defaults.

### Levels

The account data may exist at both the global and room level. The global setting defines the preference for

> all rooms, unless a per-room setting overrides it.

Author: Yeah, I kinda figured it was obvious that this setting is only really useful at the global level, but I also didn't feel like it was worth complicating the MSC by stating that the fields have different levels. 

My expectation is that implementations will just resolve the correct account for global/room each time they evaluate a piece of media and pick the first value. I hadn't included that resolving piece though. edc287ed4d5dd00b14877746241098fc0ee5e494 expands my thinking.

> all rooms, unless a per-room setting overrides it.

Reviewer: It'd be worth adding a sentence or two at the end of this section explicitly noting that "clients may not always have access to room-level account data when making a decision (e.g. upon being invited to a new room). In this case, they should only use the global account data when deciding how to present media".

> all rooms, unless a per-room setting overrides it.

Author: Noted in 3b00107d

### Notes

Homeservers MAY specify a default value ahead of time for the user, by setting a default

> value internally for the account data. The user *must* be able to mutate this value, as it's

Author: I think this is the same concern as https://github.com/matrix-org/matrix-spec-proposals/pull/4278#discussion_r2979813108?

> considered a safety feature.

Author: I'm unsure on this one. To me this is a safety setting for clients and as a user I should retain control over how my client *chooses* to render media. I would find it quite scary if my homeserver has forced any value in a safety settings control.

 Whether or not the homeserver provides the media I requested based off it's own rules is a different system to me (or rather, this feels like the wrong place to put in corporate controls).

@matrix-org/trust-safety may have opinions here

> considered a safety feature.

Reviewer 4: Emitting an @matrix-org/trust-safety opinion: we love sensible defaults, we support folks having freedom to choose a value that works for them, so we'd back keeping the MUST. The argument that servers could choose whether to provide requested media feels sound.

> considered a safety feature.

Reviewer: @Johennes in a corporate environment, I think it'd be enough that a custom client would prevent users from changing this option, despite the users technically being able to do so via an API call.

> considered a safety feature.

Author: Resolving as I don't think I can action this.

# MSC4278: Media preview controls

Matrix caters to a wide variety of networks configurations. Some networks are closed and the users have a high
degree of safety, and other networks are open to the public internet and are inherently less safe. This proposal
aims to give the user more universal control over the content they see on Matrix, starting with media previews.

Most graphical Matrix clients display media in the timeline of a room with a preview. Often these are delivered
without any consent prompt, at least by default. This presents a risk to users who may be joining the network
and do not have the same safety barriers that experienced users may have configured.

With that in mind, this MSC provides a simple switch for users and homeservers to configure a global media preview
preference.

For the purposes of this MSC, the term "undesirable content" refers to images that the user does not want to see,
be it illegal, abusive or otherwise unwanted.

## Proposal

This proposal introduces a new account data key, `m.media_preview_config`.

This key contains the following content.

```json
{
  "media_previews": "off|private|on",

>   "invite_avatars": "off|on"

Reviewer: What do you think about:

* `off` -> `hide`
* `on` -> `show`

"off" sounds to me like the avatar should not be shown. Whereas instead we expect clients to initially hide them until the user has consented to viewing them.

Author: Hm, I think the symmetry with `media_previews` is more useful than defining more enums. I agree insofar that the language isn't perfect, but I think `off`/`on` is plenty clear enough.

}
```

### `media_previews`

A "media preview" is defined as any media in a room that may be rendered automatically without explicit
user consent.

Clients SHOULD show show or hide media depending upon this setting, on at least the following content types:

 - Media message types `m.image`, `m.video`, `m.file` which all can include thumbnail information or be thumbnailed
 by the homeserver.
 - The `m.sticker` event type.
 - Inline media via `img`.
 - Any event which includes thumbnail information or may be thumbnailed by the homeserver.
 - URL Preview thumbnails

The above list is not exhaustive. Other MSCs SHOULD take this setting into account when rendering content that could

> be considered a media preview, such as [MSC4027](https://github.com/matrix-org/matrix-spec-proposals/pull/4027). 

Reviewer 5: ```suggestion
be considered a media preview, such as [MSC4027](https://github.com/matrix-org/matrix-spec-proposals/pull/4027) (Custom Images in Reactions). 
```

The one exception to the rule for now is user avatars in rooms the user has joined. For the purposes of this MSC,
by joining a room the user *has* consented to seeing the user avatars contained within. See the potential issues
section for a deeper explanation.

> ### `invite_avatars`

Reviewer: I assume we're using "off" and "on" strings for extensibility in the future, otherwise I would suggest switching this to a boolean.

If that is the case, could you say so explicitly in the document?

Author: Covered in 9dad69de

`invite_avatars` refers to any **room** avatar rendered in the client, where the client has a `membership` of `invite`.

This may typically be rendered in the form of an invite dialog, or a room list preview of a room.

The avatar may refer to:

 - The `m.room.avatar` state event.
 - The inviting user's `avatar_url` from their profile (in the case of a DM).

In *all* cases where an avatar may be rendered for a room invite, the preview MUST be controlled by this setting.

### Property value

#### `off`

The client MUST NOT show any previews for any media in affected rooms. Clients will hide the media entirely,
behind a click-to-view prompt, or some other mechanism where a user is either prevented entirely or must consent
to see the media.

Users MAY individually consent to seeing media, for example by clicking on a prompt to show a preview.
If consent is given, the client SHOULD then track that consent and show the media again in the future. The
mechanism for tracking opt-in/opt-out of media is left as an implementation detail.

> #### `private`

Reviewer 6: Non-blocking comment: I'm unenthusiastic about the use of the keyword `private` here, due to the general confusion about what exactly is a "private" room. That said, I don't have a better suggestion, and I'm happy to see you've spelt out what it means in this instance.

Author: Agreed. Maybe one day when our T&S approach is more spelled out we might have some common definitions to lean on.

Previews for media MAY be shown in "private" rooms without a prompt.
A private room is any room where:

>  - The `m.room.join_rules` state exists

Reviewer 6: It has to be said that a room with no `join_rules` event is, in fact, invite-only (cf https://github.com/matrix-org/matrix-spec/issues/657#issuecomment-2514253692), so perhaps we want to treat it accordingly.

That said, it feels like the chance of a client encountering a room where there is genuinely no `join_rules` is lower than that of a client encountering some bug where it can't see the `join_rules` for some reason. Seems like erring on the side of safety is the right call.

Might be worth making that line of reasoning explicit, though?

Author: Done in 06883081

 - The `join_rule` key of this state is `invite`, `knock`, `restricted`, or `knock_restricted`.

If any other `join_rule` is set, or cannot be determined by the client then the assumption MUST be that the

> room is public and previews MUST not be shown. Future join rules may be added to this list, but it's critical

Reviewer 6: ```suggestion
room is public and previews MUST NOT be shown. Future join rules may be added to this list, but it's critical
```

that clients adopt a safety first approach here.

> The user MAY still opt-in to media if the room is "public", and opt-out if the room is "private".

Reviewer: What do you mean by "opt-out"? Showing a "hide" button?

I now realise such an option exists in Element, but as a user of other messaging platforms where such a feature does not exist, it would be helpful to introduce that concept within the proposal (perhaps in the introduction).

Author: Clarified in bf6e1d664fd70a8e35fa4cc2bfeada62f8836e7e

Note that this setting has no effect for `invite_avatars`. Avatars can only be `off` or `on` for all invites. 
Bad actors can easily send a DM to a user (which would pass the `private` check) containing unwanted
content.

#### `on`

Media SHOULD be shown in any room without a prompt.

Users SHOULD be able to individually hide media, and this preference MUST be respected over any defaults defined in `m.media_preview_config`.

This value is the **default** setting for both properties when no value is set, to keep with the previous defaults.

### Other values

> Clients MUST treat any unknown property value in either defined property as `off`, for safety reasons.

Reviewer: Could you expand on those reasons?

```suggestion
Clients MUST treat any unknown property value in either defined property as `off`. This would
prevent, for example, a user firing up an old client and suddenly seeing media after using a
newer client that has set a newer restrictive property type.
```

Author: Done in d3a05d3b

### Levels

The account data may exist at both the global and room level. The global setting defines the preference for
all rooms, unless a per-room setting overrides it.

If `m.media_preview_config` is set at both global and room levels, the client MUST prefer using a defined value
over a default value. E.g. account data of:

```jsonc
{ // Room-level account data
  "media_previews": "on"
}
```

```jsonc
{ // Global account data
  "media_previews": "off",
  "invite_avatars": "off"
}
```

would resolve to:

```jsonc
{
  "media_previews": "on"
  "invite_avatars": "off"
}
```

### Notes

Homeservers MAY specify a default value ahead of time for the user, by setting a default

> value internally for the account data. The user *must* be able to mutate this value, as it's

Reviewer: ```suggestion
value internally for the account data. The user MUST be able to mutate this value, as it's
```

Note that this setting has no effect for `invite_avatars`. Avatars can only be `off` or `on` for all invites. 
Bad actors can easily send a DM to a user (which would pass the `private` check) containing unwanted
content.

#### `on`

Media SHOULD be shown in any room without a prompt.

Users SHOULD be able to individually hide media, and this preference MUST be respected over any defaults defined in `m.media_preview_config`.

This value is the **default** setting for both properties when no value is set, to keep with the previous defaults.

### Other values

Clients MUST treat any unknown property value in either defined property as `off`. This is to allow the feature
to be extended in the future, without unsupported clients falling back to unsafe behaviours.

### Levels

The account data may exist at both the global and room level. The global setting defines the preference for
all rooms, unless a per-room setting overrides it.

If the client doesn't have access to room level account data, such as for invites then it should just read
from the global account data.

If `m.media_preview_config` is set at both global and room levels, the client MUST prefer using a defined value

> over a default value. E.g. account data of:

Reviewer 7: Can you clarify how you know what value is a default value here?

Author: Have attempted to clarify. I think the "default" was a total misnomer

not. However, this also comes with drawbacks. It would raise the bar for all clients to expose spaces
as a feature in order to make their users safer, and in practice spaces are not fully supported either
by the ecosystem (e.g. Element X) or organisations yet.

### Room state safety flag

Rooms could expose their safeness via a state event, similar to how this proposal does for account data.
However, this would give administrators too much control over your own client experience. It would put the
responsibility on safety on room administrators, who may lack the knowledge or time for proper room configuration.

It also exposes another possible attack where users are invited to a room similar to a phishing attack, and the
room state would override their personal safety settings to deliver undesirable content.

### Trust users you share a DM with

We could seek to trust media if we share a room with a user (e.g. the `m.direct`), which might be a cheap way to
establish a basis of trust for users.

### Blurring previews

One suggestion would be to blur untrusted previews rather than keeping them as off. There are a few ways that this

> could be achieved, but each have downsides.

Reviewer 8: Technically this seems less like an alternative and more like an implementation detail of how to hide media? I'd assume any client that supports MSC2448 will simply use the sender-provided blurhashes when available instead of an ugly blank box

Author: I think there was a concern that even blurhashes are not *great* if you are being sent harmful material and an umambigious blank box was preferrable.

Aside from the fact that we have now spent the last few years training users that blurhashes are a *loading* state, so using blurhashes is likely to confuse users into thinking the application is still loading the image.
